### PR TITLE
Direct shebang execution

### PR DESCRIPTION
**Pull request by doanbaotrung**
Refactor shebang parsing logic: Extract the duplicate binfmt_script parsing logic from main.c and exec.c into a shared helper function elf_parse_shebang defined in elf.c and declared in elf.h.

---
## Summary by cubic
Enables direct execution of shebang scripts by resolving interpreters in the standalone loader and execve with bounded recursion (max 5) and consistent errno handling. Adds a shared shebang parser and a host unit test for edge cases.

- **New Features**
  - Standalone loader recursively resolves nested shebangs (max 5), updates elf_path, and prepends the interpreter and optional arg to argv (e.g., #!/bin/sh -x).

- **Refactors**
  - Consolidated parsing into elf_read_shebang and updated execve to use it, rebuild argv as [interp, opt-arg, script, argv[1:]], and mirror Linux-like errors (ENOEXEC for invalid/too-long lines or buffer limits, ELOOP when depth exceeded); added test-shebang-host with Makefile/mk/tests.mk targets.

<sup>Written for commit 10b9d69870d167c0b7185824d0183fd314a3bb1f. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/sysprog21/elfuse/pull/116?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

**Comment by jserv** [comment]
Avoid unnecessary capitalization. i.e., "Direct Shebang Execution" -> "Direct shebang execution"
See https://cbea.ms/git-commit/ carefully

File: src/core/elf.c [hunk 1]
```diff
@@ -427,3 +428,90 @@ void elf_resolve_interp(const char *sysroot,
...
+int elf_parse_shebang(const char *host_path,
+                      char *interp_out,
+                      size_t interp_sz,
+                      char *arg_out,
+                      size_t arg_sz)
+{
+    int fd = open(host_path, O_RDONLY);
+    if (fd < 0) {
+        return -errno;
+    }
```
Comment (jserv) [inline src/core/elf.c]: Rewrite as following:
```
    if (fd < 0)
        return -errno;
```

Use brackets when necessary.

File: src/core/elf.c [hunk 2]
```diff
@@ -427,3 +428,90 @@ void elf_resolve_interp(const char *sysroot,
...
+    char buf[512];
+    ssize_t nread = read(fd, buf, sizeof(buf) - 1);
+    close(fd);
+
+    if (nread < 0) {
+        return -errno;
+    }
+    if (nread < 2 || buf[0] != '#' || buf[1] != '!') {
+        return 0; /* Not a shebang script */
+    }
```
Comment (jserv) [inline src/core/elf.c:451]: It is confusing to have shebang detection in function `elf_parse_shebang`. Clarify it.

Comment (doanbaotrung) [reply src/core/elf.c:451]: That is a fair point. The function is responsible for both detection (checking for the `#!` signature) and parsing. Combining them into a single function avoids having to open, read, and close the file twice (once for detection and once for parsing).
> 
> To clarify this dual responsibility, I have:
> 1. Updated the doc comment in [elf.h](file:///Users/dbaotrung/personal/muplar/third_party/elfuse/src/core/elf.h#L139-L148) and the function header in [elf.c](file:///Users/dbaotrung/personal/muplar/third_party/elfuse/src/core/elf.c#L438-L442) to explicitly state that `elf_parse_shebang` performs both detection (returning `0` if not a shebang script) and parsing.
> 2. Kept the callers in `main.c` and `exec.c` simple by delegating both checks entirely to this function."

File: src/syscall/exec.c [hunk 3]
```diff
@@ -332,69 +332,30 @@ int64_t sys_execve(hv_vcpu_t vcpu,
...
-        char shebang_buf[256];
-        ssize_t nread = read(script_fd, shebang_buf, sizeof(shebang_buf) - 1);
-        close(script_fd);
-
-        if (nread < 2 || shebang_buf[0] != '#' || shebang_buf[1] != '!') {
-            err = -LINUX_ENOEXEC;
+        char interp_start[256];
+        char interp_arg[256];
+        int rc =
+            elf_parse_shebang(path_host, interp_start, sizeof(interp_start),
```
Comment (jserv) [inline src/syscall/exec.c]: High: the execve path resolves only one shebang layer. The interpreter re-load at exec.c:429 handles a single level and returns -LINUX_ENOENT on failure, but the interpreter file exists (path_translate already succeeded) -- so a script whose interpreter is itself a script fails with the wrong errno. main.c loops correctly to max_shebang_depth. Suggest a shared bounded resolver: loop, and return -LINUX_ENOEXEC for a bad format or -LINUX_ELOOP on depth exhaustion.

File: src/syscall/exec.c [hunk 4]
```diff
@@ -332,69 +332,30 @@ int64_t sys_execve(hv_vcpu_t vcpu,
...
 
         log_debug("execve: shebang interp=\"%s\" arg=\"%s\" script=\"%s\"",
-                  interp_start, interp_arg ? interp_arg : "(none)", path);
+                  interp_start, has_arg ? interp_arg : "(none)", path);
 
         /* Rebuild argv: [interpreter, optional-arg, script-path,
          * original-argv[1:]]
          */
-        int new_argc = 1 + (interp_arg ? 1 : 0) + 1 + (argc > 1 ? argc - 1 : 0);
+        int new_argc = 1 + (has_arg ? 1 : 0) + 1 + (argc > 1 ? argc - 1 : 0);
```
Comment (jserv) [inline src/syscall/exec.c]: Medium: new_argc is composed as int and checked against MAX_ARGS after the additions. Safe at MAX_ARGS=256 today, but check before adding and make it size-based: prefix = 2 + has_arg; reject argc > MAX_ARGS - prefix + 1.

File: src/core/elf.c [hunk 5]
```diff
@@ -427,3 +428,89 @@ void elf_resolve_interp(const char *sysroot,
     /* Strategy 3: use interp_path as-is */
     str_copy_trunc(out, interp_path, out_sz);
 }
+
+int elf_parse_shebang(const char *host_path,
```
Comment (jserv) [inline src/core/elf.c]: This is the "clarify it" point from the earlier thread. The tri-state return is fine and idiomatic (mirrors read() and binfmt handlers) -- don't split detection into a separate function that re-reads the file. Minimal fix: rename to elf_read_shebang or elf_probe_shebang so the name reflects detect-and-parse.

File: src/core/elf.c [hunk 6]
```diff
@@ -427,3 +428,89 @@ void elf_resolve_interp(const char *sysroot,
...
+                      char *interp_out,
+                      size_t interp_sz,
+                      char *arg_out,
+                      size_t arg_sz)
+{
+    /* This function handles both shebang detection and parsing. It opens the
```
Comment (jserv) [inline src/core/elf.c]: This comment restates that the function does two things; the contract already lives on the header declaration. Drop it -- if the rename lands, it is redundant.

File: src/core/elf.c [hunk 7]
```diff
@@ -427,3 +428,89 @@ void elf_resolve_interp(const char *sysroot,
...
+     * file, checks for the presence of the '#!' header, and extracts the
+     * interpreter and optional argument.
+     */
+    int fd = open(host_path, O_RDONLY);
+    if (fd < 0)
+        return -errno;
+
+    char buf[512];
```
Comment (jserv) [inline src/core/elf.c:442]: Low: reads 511 of 512 bytes then treats the first line as complete; an over-long interpreter path is silently truncated rather than rejected. str_copy_trunc's size check catches the common case. Consider rejecting an unterminated first line. (Still more generous than Linux BINPRM_BUF_SIZE=256.)

**Review by jserv — COMMENTED** [review]
Shebang refactor is a clear improvement. One correctness issue worth addressing (nested shebang via execve), plus the naming clarification raised in the earlier thread. Lower-value notes: the argv buf_off in exec.c recovers its offset via pointer subtraction from argv[argc-1] -- brittle, better returned from read_string_array; and the CR/early-EOL dialect is worth a one-line contract note plus tests (CRLF, blank interp, no trailing newline).

**Comment by jserv** [comment]
When writing Git commit messages, avoid formatting them like this:
```
Implemented a recursive shebang parsing loop
in the standalone loader's entry point.
It reads shebang interpreter lines
(including optional arguments like #!/bin/sh -x),
updates elf_path to point to the interpreter,
prepends them to guest_argv, and resolves the
host path recursively up to a depth of 5 levels.
```

At first glance, this reads more like a haiku than a commit message. Since [Git conventionally allows up to 72 characters per line in the message body](https://cbea.ms/git-commit/), wrap the text naturally while making good use of the available width. For example:
```
Implemented a recursive shebang parsing loop in the standalone loader's
entry point. It reads shebang interpreter lines, including optional
arguments like #!/bin/sh -x, updates elf_path to the interpreter, and
prepends them to guest_argv while resolving the host path recursively
up to a maximum recursion depth of five interpreter levels.
```

The goal is not to make every line exactly 72 characters long, but to avoid unnecessarily short lines. Wrap at sensible boundaries so that the text remains easy to read while making efficient use of the available line width.

**Comment by doanbaotrung** [comment]
I have implemented all the requested updates from the new code review comments:

1. **Nested Shebang Resolution in `sys_execve`**: Refactored the `execve` path in [exec.c](file:///Users/dbaotrung/personal/muplar/third_party/elfuse/src/syscall/exec.c#L330-L440) into a bounded recursive loop (resolving up to 5 shebang layers recursively). Bounded depth exhaustion returns `-LINUX_ELOOP`. Temporary materialized paths are unlinked at each iteration so they don't leak.
2. **Safe `argc` check**: Changed `new_argc` verification in [exec.c](file:///Users/dbaotrung/personal/muplar/third_party/elfuse/src/syscall/exec.c#L360) to a safer size-based condition (`argc > MAX_ARGS - prefix + 1`) that checks the boundary *before* addition, preventing potential overflow.
3. **Naming & Documentation**:
   - Renamed `elf_parse_shebang` to `elf_read_shebang` in all declarations, headers, and invocations to clarify it performs both probing (detection) and parsing.
   - Dropped the redundant description comment inside [elf.c](file:///Users/dbaotrung/personal/muplar/third_party/elfuse/src/core/elf.c#L429-L432).
4. **Overlong/Unterminated lines**: Used `strpbrk(buf + 2, "\r\n")` in [elf.c](file:///Users/dbaotrung/personal/muplar/third_party/elfuse/src/core/elf.c#L457) to locate line endings. If no line ending is found and the buffer is completely filled, it returns `-ENOEXEC`.
5. **Host-Side Unit Tests**: Added a host-side unit test file [test-shebang-host.c](file:///Users/dbaotrung/personal/muplar/third_party/elfuse/tests/test-shebang-host.c) that exercises:
   - CRLF, LF, and CR dialects.
   - Scripts ending without trailing newlines (EOF).
   - Empty/blank shebang interpreters.
   - Non-shebang scripts.
   - Overlong/unterminated shebang lines.
   - The test was successfully linked and runs automatically with `make check`.

All changes compile cleanly and pass the native unit tests.

**Comment by jserv** [comment]
> All changes compile cleanly and pass the native unit tests. You can review the details in walkthrough.md.

We are humans. Don't copy agent-specific reply here.

**Comment by doanbaotrung** [comment]
Got it sir

**Comment by doanbaotrung** [comment]
Do you have any tool to help format the git commit message?
I break line manually so some line longer than others and not well formatted

**Comment by jserv** [comment]
Thank @doanbaotrung for contributing!
